### DISCUSSION TITLE
pytest structure for error handler tests

### DISCUSSION THREAD

**Discussion by pavlis** [mspass-team/mspass#452] [Design & Development]
In working on revision 2 of Database I've had to walk through a lot of lines of pytest code.   I have repeatedly had to deal with pytest sections like the following:
```
# test doc that does not exist
            non_exist_id = ObjectId()
            with pytest.raises(
                MsPASSError,
                match="Database.verify:  objectid="
                + str(non_exist_id)
                + " has no matching document in wf_TimeSeries",
            ):
                problematic_keys = self.db.verify(
                    non_exist_id, "wf_TimeSeries", tests=["xref", "type", "undefined"]
                )
```
The problem I have with this structure is it is fragile in the sense that someone might want to change an obscure error message to expand or rephrase the error message.   My key point is the test should avoid explicit testing of the text of the error message.  It should normally be enough to just test the type of the thrown exception.   That will reduce the incidence of false test failures like that I've had to endure the last week after extensive revisions to error messages to try to make them less opaque.  

I personally find "with" usage here a bit confusing as it doesn't make it as obvious from the context this is a test of an error handler.   I would recommend the above be rewritten like this:
```
try: 
  problematic_keys = self.db.verify(
                    non_exist_id, "wf_TimeSeries", tests=["xref", "type", "undefined"]
                )
except MsPASSError as e:
  assert e.severity==ErrorSeverity.Invalid
  # alternatively just a "pass" statement
```

**Comment by wangyinz** [comment]
Well, it is actually not a good idea to use the try clause in the test. This is because pytest will not be able to tell what it is testing for in the output, and when the test fails the error will be confused by the normal python exception messages. It is actually a requirement to use the `pytest.raises` to test for any exceptions.

The particular problem you ran into is actually due to the wrong usage of the match clause. In `pytest.raises` here, the match here actually means partial match instead of exact match. This means, you can just have the keyword in the match clause that you know won't be change in the error message to pass the test. You can see in the tests I previously wrote how it should be used. For example:
https://github.com/mspass-team/mspass/blob/23898f45753603606c529b0034e3bac7eaf80ac2/python/tests/test_ccore.py#L582-L583

You can see that it seeks to match "not recognized", which means as long as there is a substring of "not recognized" in the error message this test will pass. This way, it is actually more robust and powerful than just using the try clause. Also, I found the following example that demonstrate how the match keyword is used: 
```
import pytest

def test_exception():
    with pytest.raises(ValueError, match="invalid value"):
        raise ValueError("This is an invalid value error")

    with pytest.raises(ValueError, match="invalid"):
        raise ValueError("This is an invalid value error")

    with pytest.raises(ValueError, match="error"):
        raise ValueError("This is an invalid value error")

    with pytest.raises(ValueError, match="value error"):
        raise ValueError("This is an invalid value error")
```
In this example, all four test will pass because of the partial matching. 

So, it is pretty clear that in the code you quoted, it should really just have something like `match="no matching"` in it instead of that long complicated string.

**Comment by pavlis** [comment]
Thanks for that education.   I will modify what I find to use more brief strings that are unlikely to change.   All future MsPASS development should recognize: (a) the need for the "with" clause, and (b) the need to not use the full message but a few keywords that are unlikely to change.
